### PR TITLE
enable autocast for Lazy CUDA device

### PR DESCRIPTION
**Pull request by jjsjann123**
This enables the autocast for lazy device similar to what XLA does

Note that this isn't a safe approach, as we default to AutocastCUDA here, but Lazy fallback could be either a CPU or CUDA.

### CODE REVIEW

**Comment by wconstab** [comment]
we do have a new lazy backend API for getting the eager fallback device.  we could call that if the tensor is lazy and determine if the backend eager fallback will be CPU/CUDA.

https://github.com/pytorch/pytorch/blob/lazy_tensor_staging/torch/csrc/lazy/backend/backend_interface.h#L85

**Comment by jjsjann123** [comment]
Using this: https://github.com/pytorch/pytorch/pull/69090/commits/0eb4c9c4e737699d65f339123202f84b15c83d4c
throws me linker errors building tests. Any chance I can get some help from someone who knows the build dependencies?
```
/usr/bin/ld: lib/libc10.so: undefined reference to `torch::lazy::getBackend()'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by wconstab** [comment]
> Using this: [0eb4c9c](https://github.com/pytorch/pytorch/commit/0eb4c9c4e737699d65f339123202f84b15c83d4c) throws me linker errors building tests. Any chance I can get some help from someone who knows the build dependencies?
> 
> ```
> /usr/bin/ld: lib/libc10.so: undefined reference to `torch::lazy::getBackend()'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```

@alanwaketan found and fixed an issue with the backend registrar earlier this week.  I am not sure if it's the same issue you're facing, but there shouldn't be any particular tricks with build dependencies other than building pytorch first and then lazy tensor. Can you try pulling our latest staging branch?

**Comment by jjsjann123** [comment]
> 

Unfortunately I'm already up-to-date. :cry:

**Comment by alanwaketan** [comment]
> Unfortunately I'm already up-to-date. 😢
> 
> Maybe it's a stale build issue, let me do a clean up and rebuild

You can ping me in Slack if you need more helps!

**Comment by malfet** [comment]
> 



> Using this: [0eb4c9c](https://github.com/pytorch/pytorch/commit/0eb4c9c4e737699d65f339123202f84b15c83d4c) throws me linker errors building tests. Any chance I can get some help from someone who knows the build dependencies?
> 
> ```
> /usr/bin/ld: lib/libc10.so: undefined reference to `torch::lazy::getBackend()'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```
> 
> FYI, building from scratch with that changes gives missing headers:
> 
> ```
> [60/1837] Building CXX object c10/CMakeFiles/c10.dir/core/DispatchKeySet.cpp.o
> FAILED: c10/CMakeFiles/c10.dir/core/DispatchKeySet.cpp.o
> /usr/bin/clang++-8 -DFMT_HEADER_ONLY=1 -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMAGMA_V2 -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DSUPPORTS_BACKTRACE=1 -DTH_BLAS_MKL -DUSE_EXTERNAL_MZCRC -D_FILE_OFFSET_BITS=64 -Dc10_EXPORTS -Iaten/src -I../aten/src -I. -I../ -I../cmake/../third_party/benchmark/include -I../cmake/../third_party/cudnn_frontend/include -I../third_party/onnx -Ithird_party/onnx -I../third_party/foxi -Ithird_party/foxi -I../c10/.. -I../third_party/fmt/include -isystem third_party/gloo -isystem ../cmake/../third_party/gloo -isystem ../cmake/../third_party/googletest/googlemock/include -isystem ../cmake/../third_party/googletest/googletest/include -isystem ../third_party/protobuf/src -isystem /opt/conda/include -isystem ../third_party/gemmlowp -isystem ../third_party/neon2sse -isystem ../third_party/XNNPACK/include -isystem ../third_party -isystem ../cmake/../third_party/eigen -isystem /opt/conda/include/python3.8 -isystem /opt/conda/lib/python3.8/site-packages/numpy/core/include -isystem ../cmake/../third_party/pybind11/include -isystem /usr/local/mpi/include/openmpi -isystem /usr/local/mpi/include/openmpi/opal/mca/hwloc/hwloc201/hwloc/include -isystem /usr/local/mpi/include/openmpi/opal/mca/event/libevent2022/libevent -isystem /usr/local/mpi/include/openmpi/opal/mca/event/libevent2022/libevent/include -isystem /usr/local/mpi/include -isystem /usr/local/cuda/include -isystem ../third_party/ideep/mkl-dnn/include -isystem ../third_party/ideep/include -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp=libomp -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-typedef-redefinition -Wno-unknown-warning-option -Wno-unused-private-field -Wno-inconsistent-missing-override -Wno-aligned-allocation-unavailable -Wno-c++14-extensions -Wno-constexpr-not-const -Wno-missing-braces -Qunused-arguments -fcolor-diagnostics -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -g -fno-omit-frame-pointer -O0 -fPIC -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -DC10_BUILD_MAIN_LIB -fvisibility=hidden -std=gnu++14 -MD -MT c10/CMakeFiles/c10.dir/core/DispatchKeySet.cpp.o -MF c10/CMakeFiles/c10.dir/core/DispatchKeySet.cpp.o.d -o c10/CMakeFiles/c10.dir/core/DispatchKeySet.cpp.o -c ../c10/core/DispatchKeySet.cpp
> In file included from ../c10/core/DispatchKeySet.cpp:2:
> In file included from ../torch/csrc/lazy/backend/backend_interface.h:7:
> In file included from ../torch/csrc/lazy/backend/lowering_context.h:11:
> In file included from ../torch/csrc/lazy/core/ir.h:16:
> In file included from ../torch/csrc/lazy/core/hash.h:12:
> ../aten/src/ATen/Tensor.h:3:10: fatal error: 'ATen/core/TensorBody.h' file not found
> ```

Both failures are symptoms of the same problem: library dependencies must be a DAG, but this PR tries to introduce a cycle: `torch_cpu` depends on `c10`, so `c10` can not depend on any symbols (nor headers, though technically there is a workaround for that,  from `torch_cpu`.

File: c10/core/DispatchKeySet.cpp [hunk 1]
```diff
@@ -103,6 +104,15 @@ DispatchKeySet getAutocastRelatedKeySetFromBackend(DispatchKey t) {
     case DispatchKey::CUDA:
     case DispatchKey::XLA:
       return DispatchKeySet(DispatchKey::AutocastCUDA);
+    case DispatchKey::Lazy:
```
Comment (wconstab) [inline c10/core/DispatchKeySet.cpp:107]: I don't think we can fix this properly until we redesign lazy tensor as a 'mode'.  That said, you could work around for now by checking `LTC_TS_CUDA` env here.

**Review by wconstab — APPROVED** [review]
lgtm, thanks!

**Pull request merged**